compiler/noirc_evaluator/src/acir/generated_acir.rs

```diff
@@ -534,36 +534,32 @@ impl<F: AcirField> GeneratedAcir<F> {
     /// By setting `z` to be `0`, we can make `y` equal to `1`.
     /// This is easily observed: `y = 1 - t * 0`
     /// Now since `y` is one, this means that `t` needs to be zero, or else `y * t == 0` will fail.
-    fn is_zero(&mut self, t_expr: &Expression<F>) -> Witness {
-        // We're checking for equality with zero so we can negate the expression without changing the result.
-        // This is useful as it will sometimes allow us to simplify an expression down to a witness.
-        let t_witness = if let Some(witness) = t_expr.to_witness() {
-            witness
+    fn is_zero(&mut self, t_expr: Expression<F>) -> Witness {
+        // We're going to be multiplying this expression by two different witnesses in a second so we want to
+        // ensure that this expression only contains a single witness. We can tolerate coefficients and constant terms however.
+        let linear_t = if t_expr.is_degree_one_univariate() {
+            t_expr
         } else {
-            let negated_expr = t_expr * -F::one();
-            self.get_or_create_witness(&negated_expr)
+            Expression::<F>::from(self.get_or_create_witness(&t_expr))
         };
```

Comment:
I'm trying to understand where where exactly the generation of a new witness is avoided. 

As I see it `t_expr.to_witness()` just unwrapped the witness from `t_expr` if it was `is_degree_one_univariate`, but if it's not, then now instead of creating a witness for `-t_expr`, we create a witness for `t_expr` itself. So the number of witnesses generated so far is the same, isn't it?

Reply:
Previously we'd handle the cases `t_expr = x` and `t_expr = -x` only as we were looking to extract the witness `x`. Now we tolerate `t_expr = m*x +c` so we can avoid creating witnesses in more cases.

We can do this as the scaling factor doesn't affect the following constraints and constants in`t_expr` introduce another linear term to the following constraints which fits inside the existing gates.

Reply:
Ah because now this is from `to_witness` is not happening:
```rust
            if coefficient.is_one() && constant.is_zero() {
                return Some(variable);
            }
```